packages/lib-core/src/extensions/actions.ts

```diff
@@ -0,0 +1,119 @@
+import type { ReactNode } from 'react';
+import type { ExtensionHook, ExtensionK8sKindVersionModel } from '../api/common-types';
+import type { CodeRef, Extension } from '../types/extension';
+import type { AccessReviewResourceAttributes } from './core-types';
+
+/** ActionProvider contributes a hook that returns list of actions for specific context */
+export type ActionProvider = Extension<
+  'core.action/provider',
+  {
+    /** The context ID helps to narrow the scope of contributed actions to a particular area of the application. Ex - topology, helm */
+    contextId: string | 'resource';
+    /** A react hook which returns actions for the given scope.
+     * If contextId = `resource` then the scope will always be a K8s resource object
+     * */
+    provider: CodeRef<ExtensionHook<Action[]>>;
+  }
+>;
+
+/** ResourceActionProvider contributes a hook that returns list of actions for specific resource model */
+export type ResourceActionProvider = Extension<
+  'core.action/resource-provider',
+  {
+    /** The model for which this provider provides actions for. */
+    model: ExtensionK8sKindVersionModel;
+    /** A react hook which returns actions for the given resource model */
+    provider: CodeRef<ExtensionHook<Action[]>>;
+  }
+>;
+
+/** ActionGroup contributes an action group that can also be a submenu */
+export type ActionGroup = Extension<'core.action/group', Group>;
+
+/** ActionFilter can be used to filter an action */
+export type ActionFilter = Extension<
+  'core.action/filter',
+  {
+    /** The context ID helps to narrow the scope of contributed actions to a particular area of the application. Ex - topology, helm */
+    contextId: string | 'resource';
+    /** A function which will filter actions based on some conditions.
+     * scope: The scope in which actions should be provided for.
+     * Note: hook may be required if we want to remove the ModifyCount action from a deployment with HPA
+     * */
+    filter: CodeRef<(scope: unknown, action: Action) => boolean>;
+  }
+>;
+
+// Type guards
+
+export const isActionProvider = (e: Extension): e is ActionProvider => {
+  return e.type === 'core.action/provider';
+};
+
+export const isResourceActionProvider = (e: Extension): e is ResourceActionProvider => {
+  return e.type === 'core.action/resource-provider';
+};
+
+export const isActionGroup = (e: Extension): e is ActionGroup => {
+  return e.type === 'core.action/group';
+};
+
+export const isActionFilter = (e: Extension): e is ActionFilter => {
+  return e.type === 'core.action/filter';
+};
+
+// Arbitrary types
+
+export type Group = {
```

Comment:
By convention, these are helper types which are relevant for / used in this file only.